File: 383-ransom-note/383-ransom-note.py
class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        magazine_count = collections.Counter(magazine)
        ransom_count = collections.Counter(ransomNote)
        
        for i in range(26):
            cur = chr(i + ord('a'))
            x1 = magazine_count.setdefault(cur, 0)
            x2 = ransom_count.setdefault(cur, 0)
           # print(x1, x2, cur)
            if x1 < x2:
                return False
        return True

File: 383-ransom-note/383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> counter = new HashMap<>();
        int N = magazine.length();
        for(int i = 0; i < N; i++){
            char cur_char = magazine.charAt(i);
            if(counter.containsKey(cur_char)){
                
                int cur_val = counter.get(cur_char);
                counter.put(cur_char, cur_val+1);
            }else{
                counter.put(cur_char, 1);
            }
        }
        for(int j = 0; j < ransomNote.length(); j++){
            char cur_char = ransomNote.charAt(j);
            if (counter.containsKey(cur_char)){
                if (counter.get(cur_char) == 0){
                    return false;
                }
                int cur_val = counter.get(cur_char);
                counter.put(cur_char, cur_val-1);
            }else{
                return false;
            }
        }
        return true;
        
    }
}